File: prolog/battle_sim.pl
% Wicked Zerg - Battle Simulation
% Phase 128: Prolog

battle_sim(Units, Damage) :-
    length(Units, Count),
    Damage is Count * 5.

calculate_swarm_damage(Count, Damage) :-
    Damage is Count * 5.

swarm_formation(CenterX, CenterY, Count, Radius, Positions) :-
    findall((X, Y),
            (between(0, Count, I),
             Angle is 2 * pi * I / Count,
             X is CenterX + Radius * cos(Angle),
             Y is CenterY + Radius * sin(Angle)),
            Positions).

unit_strength(Health, Damage, Armor, Strength) :-
    Effective is Damage * Health / 100,
    Strength is Effective * (1 - Armor * 0.01).

battle_outcome(Attackers, Defenders, AttackerWins) :-
    maplist(unit_strength, Attackers, AttackPowers),
    maplist(unit_strength, Defenders, DefensePowers),
    sumlist(AttackPowers, TotalAttack),
    sumlist(DefensePowers, TotalDefense),
    AttackerWins is TotalAttack > TotalDefense.

distance((X1, Y1), (X2, Y2), Dist) :-
    Dist is sqrt((X1 - X2) ^ 2 + (Y1 - Y2) ^ 2).

min_distance(Point, Points, MinDist) :-
    maplist(distance(Point), Points, Dists),
    min_list(Dists, MinDist).

calculate_threats(OurPositions, EnemyPositions, ThreatCount) :-
    maplist(min_distance(OurPositions), EnemyPositions, MinDists),
    include(<(10), MinDists, Threats),
    length(Threats, ThreatCount).

File: pascal/BattleSim.pas
program BattleSim;
{$mode objfpc}{$H+}

uses
  Math;

type
  TBattleUnit = record
    UnitType: Integer;
    Health: Real;
    Damage: Real;
    Armor: Real;
    PosX: Real;
    PosY: Real;
  end;

function CalculateSwarmDamage(Count: Integer): Integer;
begin
  Result := Count * 5;
end;

procedure SwarmFormation(CenterX, CenterY: Real; Count: Integer; Radius: Real;
  var Positions: array of Real);
var
  I: Integer;
  Angle: Real;
begin
  for I := 0 to Count - 1 do
  begin
    Angle := 2 * Pi * I / Count;
    Positions[I * 2] := CenterX + Radius * Cos(Angle);
    Positions[I * 2 + 1] := CenterY + Radius * Sin(Angle);
  end;
end;

function UnitStrength(Health, Damage, Armor: Real): Real;
var
  Effective: Real;
begin
  Effective := Damage * Health / 100;
  Result := Effective * (1 - Armor * 0.01);
end;

function BattleOutcome(Attackers, Defenders: array of Real): Boolean;
var
  AttackPower, DefensePower: Real;
  I: Integer;
begin
  AttackPower := 0;
  DefensePower := 0;
  for I := 0 to High(Attackers) div 3 do
    AttackPower := AttackPower + UnitStrength(Attackers[I*3], Attackers[I*3+1], Attackers[I*3+2]);
  for I := 0 to High(Defenders) div 3 do
    DefensePower := DefensePower + UnitStrength(Defenders[I*3], Defenders[I*3+1], Defenders[I*3+2]);
  Result := AttackPower > DefensePower;
end;

begin
  WriteLn('Battle Simulation Initialized - Pascal');
end.
